File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int bal=0;
        int i=1,j=prices.length-1;
        int min=prices[0],max=0;
        
        while(i<=j){
            int cur = prices[i]-min;
            max = Math.max(max,cur);
            min = Math.min(min,prices[i]);
            i++;    
        }
        return max;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxi=prices[0]
        mini=prices[0]
        pro=0
        for i in prices[1:]:
            if i < mini:
                pro= max(maxi-mini,pro)
                mini=i
                maxi=i
            if i> maxi:
                maxi=i
        return max(pro,maxi-mini)